ngraph/core/src/preprocess/preprocess_steps_impl.cpp

```diff
@@ -0,0 +1,94 @@
+
+#include "ngraph/opsets/opset1.hpp"
+#include "openvino/core/node.hpp"
+#include "openvino/core/shape.hpp"
+
+using namespace ov;
+using namespace ov::preprocess;
+
+static Shape construct_mean_scale_shape(const std::shared_ptr<Node>& node,
+                                        size_t values_size,
+                                        const PreprocessingContext& context) {
+    // TODO: support also Mean/Scale image case
+    OPENVINO_ASSERT(layout::has_channels(context.layout()), "Channels dimension is not specified in layout");
+    auto channels_index = layout::channels(context.layout());
+    auto node_shape = node->get_output_partial_shape(0);
+    auto node_rank = node->get_output_partial_shape(0).rank();
+    OPENVINO_ASSERT(node_rank.is_static(), "Mean/scale vector operation is not supported for fully dynamic shape");
+    std::vector<std::size_t> v(node_rank.get_length(), 1);
+    if (channels_index < 0) {
+        // E.g. channels_index = -1 means last dimension
+        channels_index = node_rank.get_length() + channels_index;
+    }
+    OPENVINO_ASSERT(node_rank.get_length() > channels_index, "Channels dimension is out of bounds");
+    OPENVINO_ASSERT(node_shape[channels_index] == values_size, "Number of channels and mean/values size mismatch");
+    v[channels_index] = values_size;
+    return {v};
+}
+
+void PreProcessSteps::PreProcessStepsImpl::add_scale_impl(const std::vector<float>& values) {
+    m_actions.emplace_back(std::make_tuple(
+        [values](const std::vector<std::shared_ptr<Node>>& nodes, PreprocessingContext& context) {
+            OPENVINO_ASSERT(!nodes.empty(), "Internal error: Can't apply scale preprocessing for empty input.");
+            OPENVINO_ASSERT(nodes.size() == 1,
+                            "Can't apply scale preprocessing for multi-plane input. Suggesting to convert current "
+                            "image to RGB/BGR color format using 'convert_color'");
+            Shape shape;
+            if (values.size() == 1) {
+                shape = Shape{1};
+            } else {
+                shape = construct_mean_scale_shape(nodes[0], values.size(), context);
+            }
+            auto constant = op::v0::Constant::create(element::f32, shape, values);
+            constant->set_friendly_name(nodes[0]->get_friendly_name() + "/scale/Divide_Factor");
+
+            auto new_op = std::make_shared<op::v1::Divide>(nodes[0], constant);
+            new_op->set_friendly_name(nodes[0]->get_friendly_name() + "/scale/Divide");
+            return new_op;
+        },
+        false));
+}
+
+void PreProcessSteps::PreProcessStepsImpl::add_mean_impl(const std::vector<float>& values) {
+    m_actions.emplace_back(std::make_tuple(
+        [values](const std::vector<std::shared_ptr<Node>>& nodes, PreprocessingContext& context) {
+            OPENVINO_ASSERT(!nodes.empty(), "Internal error: Can't apply mean preprocessing for empty input.");
+            OPENVINO_ASSERT(nodes.size() == 1,
```

Comment:
may be once we merge NV12, I420 support we need to check based on color format instead of number of inputs? because NV12_SINGLE_PLANE seems not applicable for mean image as well.. Or at least it looks very strange because usually mean / scale are applied before the network input

Reply:
ok, i'll add this check, but I thought that 'mean/scale' can be applied to any tensor if user wants to do this, so with layout 'NC' we can also apply mean/scale even if this is definitely not an image.
I can add additional check `if (color_format != UNDEFINED && !is_rgb_family(color_format))` - will it cover your needs?

Reply:
right